Keepers that never become master/sync replica

rnaveiras opened the pull request:
Implements the proposal https://github.com/sorintlab/stolon/issues/696

sgotti commented on cmd/keeper/cmd/keeper.go:
```diff
@@ -114,6 +114,9 @@ type config struct {
 	pgInitialSUPasswordFile string
+
+	neverMaster             bool
+	neverSynchronousReplica bool
```
Just a poll... Since I don't lilke negated flags. What about `canBeMaster` `canBeSynchronousReplica` that defaults to true?

lawrencejones replied:
I think you're right about that, negative flags aren't usually the most user friendly. We originally chose this as it felt natural to specify that you couldn't become a particular role, but I don't think that makes sense on second thought. 

Let's go with your suggestion!

sgotti commented on cmd/keeper/cmd/keeper.go:
```diff
@@ -456,6 +462,9 @@ type PostgresKeeper struct {
 	waitSyncStandbysSynced bool
+
+	neverMaster             bool
+	neverSynchronousReplica bool
```
Same here (since I don't lilke negated flags). What about `canBeMaster` `canBeSynchronousReplica` that defaults to true? The unique issue is that if you update the sentinel but not the keeper it'll have `canBeMaster` as false so we should make it a `*bool` and default it to true when nil.

sgotti commented:
@rnaveiras Is this ready for review? Can you please rebase?

rnaveiras commented:
@sgotti I just rebased and fixed the conflicts. It's ready for review. I'm happy to squash the commits if you think is necessary after you finish your review.

sgotti commented on cmd/sentinel/cmd/sentinel_test.go:
```diff
@@ -4960,6 +4961,394 @@ func TestUpdateCluster(t *testing.T) {
+						Status: cluster.KeeperStatus{
+							Healthy:                 true,
+							LastHealthyTime:         now,
+							CanBeSynchronousReplica: newBool(false),
```
cluster.BoolP(false)

sgotti commented on cmd/sentinel/cmd/sentinel_test.go:
```diff
@@ -5103,3 +5493,7 @@ func testEqualCD(cd1, cd2 *cluster.ClusterData) bool {
 
 }
+
+func newBool(b bool) *bool {
```
Unneeded since there already cluster.BoolP

leev commented:
Any chance this could be rebased and merged soon?  I was working on this feature tonight and when coming to submit a PR, I only found this then.

rnaveiras commented:
> Any chance this could be rebased and merged soon? I was working on this feature tonight and when coming to submit a PR, I only found this then.

Hi @leev, yes I will take care of make the changes tomorrow. Sorry for the delay, it's been in my to-do list for months. Thanks for bringing it up.

rnaveiras commented:
@sgotti I'm happy to squash the commits if you thing is worth it after the final review. Is possible for me to trigger again the CI e2e tests? They're failing now after the rebase.

sgotti commented:
@rnaveiras It LGTM. Yes, please squash in a single commit. Thanks!

Currently the run.agola.io instance configuration doesn't permit restart from the UI if not authenticated (and registration is currently disabled). I'll restart it for you if I noticed it failed just for some flakes (since agola could also restart only failed tasks). If you want to trigger a full restart you could amend your commit just to generate a new commit id and force push.

rnaveiras commented:
> @rnaveiras It LGTM. Yes, please squash in a single commit. Thanks!

Fantastic!! I just squash all commits. 

> Currently the run.agola.io instance configuration doesn't permit restart from the UI if not authenticated (and registration is currently disabled). I'll restart it for you if I noticed it failed just for some flakes (since agola could also restart only failed tasks). If you want to trigger a full restart you could amend your commit just to generate a new commit id and force push.

Thanks!

rnaveiras commented:
@sgotti all CI checks are passing now!!

sgotti commented:
@rnaveiras Thanks! Merging.

leev commented:
Nice!  Thank you.
